### PR TITLE
Add Typing to Llama Training

### LINKED ISSUE #4: Add typing to train_llama.py
- Currently, we put #type: ignore at the top of the train_llama.py file.
- This should be remove and all typing errors should be fixed
- Using jaxtyping (e.g. Float[Tensor, "batch seq d_embd"]) is desired, even if it's purely for documentation and not actually used in runtime checking (I've found runtime checking typically makes the documentation a little worse).

### PR DESCRIPTION
**Pull request by lennart-finke**
## Description
Added Jax and regular typing to `llama_train.py`

## Related Issue
Should close #4.

## How Has This Been Tested?
Executing the script in a Colab instance.

### CODE REVIEW

**Comment by lennart-finke** [comment]
Alright, will have to work on this some more as the checks indicate.

**Comment by lennart-finke** [comment]
Now it might go through. If someone has time, I recommend double checking the jaxtyping hints though, as I was not entirely sure everywhere.

File: simple_stories_train/train_llama.py [hunk 1]
```diff
@@ -152,9 +130,10 @@ def calculate_sin_cos_rotary(
         return torch.sin(angles).to(dtype), torch.cos(angles).to(dtype)
 
     def get_offset_position_ids(
+        self,
         past_kv_pos_offset: int,
-        attention_mask: torch.Tensor,
-    ) -> torch.Tensor:
+        attention_mask: Int[torch.Tensor, "batch offset_pos"],
```
Comment (danbraunai) [inline simple_stories_train/train_llama.py:135]: This will probably be Bool and not Int?

Comment (danbraunai) [inline simple_stories_train/train_llama.py:135]: Probably nicer to do `from torch import Tensor` at the top and then make all of these a bit shorter.

File: simple_stories_train/train_llama.py [hunk 2]
```diff
@@ -188,10 +167,10 @@ def rotate_every_two(self, x: torch.Tensor) -> torch.Tensor:
 
     def apply_rotary(
         self,
-        x: torch.Tensor,
+        x: Float[torch.Tensor, "batch head pos head_size"],
         past_kv_pos_offset=0,
-        attention_mask: torch.Tensor = None,
-    ) -> torch.Tensor:
+        attention_mask: None | Int[torch.Tensor, "batch offset_pos"] = None,
```
Comment (danbraunai) [inline simple_stories_train/train_llama.py:172]: Nit: I think standard to have the None after the non-none stuff, as you have later

File: simple_stories_train/train_llama.py [hunk 3]
```diff
@@ -260,28 +243,28 @@ def forward(self, x: torch.Tensor, attention_mask: torch.Tensor = None) -> torch
...
         self.act_fn = nn.functional.silu
 
-    def forward(self, x):
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
         return self.down_proj(self.act_fn(self.gate_proj(x)) * self.up_proj(x))
 
 
 class LlamaRMSNorm(nn.Module):
-    def __init__(self, hidden_size, eps=1e-6):
+    def __init__(self, hidden_size: int, eps=1e-6):
```
Comment (danbraunai) [inline simple_stories_train/train_llama.py:259]: nit: type `eps` to float

File: simple_stories_train/train_llama.py [hunk 4]
```diff
@@ -260,28 +243,28 @@ def forward(self, x: torch.Tensor, attention_mask: torch.Tensor = None) -> torch
...
+    def __init__(self, config: LlamaConfig):
         super().__init__()
         self.config = config
         self.gate_proj = nn.Linear(config.n_embd, config.n_intermediate, bias=config.mlp_bias)
         self.up_proj = nn.Linear(config.n_embd, config.n_intermediate, bias=config.mlp_bias)
         self.down_proj = nn.Linear(config.n_intermediate, config.n_embd, bias=config.mlp_bias)
         self.act_fn = nn.functional.silu
 
-    def forward(self, x):
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
```
Comment (danbraunai) [inline simple_stories_train/train_llama.py:254]: Probably best to type as `def forward(self, x: Float[Tensor, "... dim"]) -> Float[Tensor, "... dim"]`.

File: simple_stories_train/train_llama.py [hunk 5]
```diff
@@ -260,28 +243,28 @@ def forward(self, x: torch.Tensor, attention_mask: torch.Tensor = None) -> torch
...
+    def __init__(self, hidden_size: int, eps=1e-6):
         """
         LlamaRMSNorm is equivalent to T5LayerNorm
         """
         super().__init__()
         self.weight = nn.Parameter(torch.ones(hidden_size))
         self.variance_epsilon = eps
 
-    def forward(self, hidden_states):
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
```
Comment (danbraunai) [inline simple_stories_train/train_llama.py:267]: Can use `def forward(self, hidden_states: Float[Tensor, "... dim"]) -> Float[Tensor, "... dim"]`.

File: simple_stories_train/train_llama.py [hunk 6]
```diff
@@ -293,25 +276,23 @@ def extra_repr(self):
...
         super().__init__()
         self.rms_1 = LlamaRMSNorm(config.n_embd)
         self.attn = CausalSelfAttention(config)
         self.rms_2 = LlamaRMSNorm(config.n_embd)
         self.mlp = SwiGLUMLP(config)
 
-    def forward(self, x):
+    def forward(
+        self, x: Float[torch.Tensor, "batch pos d_model"]
+    ) -> Float[torch.Tensor, "batch pos d_model"]:
```
Comment (danbraunai) [inline simple_stories_train/train_llama.py:288]: You can actually type as 
```
    def forward(
        self, x: Float[torch.Tensor, "... pos d_model"]
    ) -> Float[torch.Tensor, "... pos d_model"]:
```
because this will work whether there is a batch dimension or not (or e.g. two batch dimensions). "..." stands for zero or more axes https://docs.kidger.site/jaxtyping/api/array/

File: simple_stories_train/train_llama.py [hunk 7]
```diff
@@ -323,7 +304,7 @@ def __init__(self, config):
             )
         )
         self.lm_head = nn.Linear(config.n_embd, config.vocab_size, bias=False)
-        self.lm_head.LLMC_SKIP_INIT = 1  # don't init this one, we will tie weights
+        self.LLMC_SKIP_INIT = 1  # don't init this one, we will tie weights
```
Comment (danbraunai) [inline simple_stories_train/train_llama.py:307]: This change seems sus

File: simple_stories_train/train_llama.py [hunk 8]
```diff
@@ -333,24 +314,29 @@ def __init__(self, config):
...
         self.apply(self._init_weights)
 
-    def _init_weights(self, module):
+    def _init_weights(self, module: nn.Module):
         if isinstance(module, nn.Linear):
             # apply special scaled init to the residual projections, per GPT-2 paper
             std = (
                 0.02
-                if not hasattr(module, "LLMC_RESIDUAL_SCALE_FLAG")
+                if not hasattr(self, "LLMC_RESIDUAL_SCALE_FLAG")
```
Comment (danbraunai) [inline simple_stories_train/train_llama.py:322]: doesn't seem right. The self.apply will apply to every submodule.

File: simple_stories_train/train_llama.py [hunk 9]
```diff
@@ -333,24 +314,29 @@ def __init__(self, config):
...
+            if module.bias is not None:  # type: ignore
                 torch.nn.init.zeros_(module.bias)
         elif isinstance(module, nn.Embedding):
             torch.nn.init.normal_(module.weight, mean=0.0, std=0.02, generator=self.init_rng)
 
-    def forward(self, idx, targets=None, return_logits=True):
+    def forward(
+        self,
+        idx: Float[torch.Tensor, "batch pos"],
+        targets: None | Float[torch.Tensor, "batch pos vocab"] = None,
```
Comment (danbraunai) [inline simple_stories_train/train_llama.py:337]: nit: None after other type. And Float[torch.Tensor, "... pos vocab"]

File: simple_stories_train/train_llama.py [hunk 10]
```diff
@@ -333,24 +314,29 @@ def __init__(self, config):
...
+        return_logits=True,
+    ):
```
Comment (danbraunai) [inline simple_stories_train/train_llama.py:339]: return annotation missing

File: simple_stories_train/train_llama.py [hunk 11]
```diff
@@ -446,7 +441,14 @@ def from_pretrained(cls, model_type):
...
 
-    def configure_optimizers(self, weight_decay, learning_rate, betas, device_type, zero_stage):
+    def configure_optimizers(
+        self,
+        weight_decay: float,
+        learning_rate: float,
+        betas: tuple[float, float],
+        device_type: str,
+        zero_stage: int,
+    ):
```
Comment (danbraunai) [inline simple_stories_train/train_llama.py:451]: missing return type

File: simple_stories_train/train_llama.py [hunk 12]
```diff
@@ -489,7 +492,13 @@ def configure_optimizers(self, weight_decay, learning_rate, betas, device_type,
...
 
     @torch.no_grad()
-    def generate(self, idx, max_new_tokens, temperature=1.0, top_k=None):
+    def generate(
+        self,
+        idx: Float[torch.Tensor, "batch pos"],
+        max_new_tokens: int,
+        temperature=1.0,
+        top_k: None | int = None,
+    ) -> Float[torch.Tensor, "batch pos"]:
```
Comment (danbraunai) [inline simple_stories_train/train_llama.py:501]: probably works with "... pos"?

Nit: None before int.

File: simple_stories_train/train_llama.py [hunk 13]
```diff
@@ -522,7 +531,7 @@ def generate(self, idx, max_new_tokens, temperature=1.0, top_k=None):
 # Our own simple Distributed Data Loader
 
 
-def _peek_data_shard(filename):
+def _peek_data_shard(filename: str):
```
Comment (danbraunai) [inline simple_stories_train/train_llama.py:534]: missing return type

File: simple_stories_train/train_llama.py [hunk 14]
```diff
@@ -540,7 +549,7 @@ def _peek_data_shard(filename):
     return ntok  # for now just return the number of tokens
 
 
-def _load_data_shard(filename):
+def _load_data_shard(filename: str):
```
Comment (danbraunai) [inline simple_stories_train/train_llama.py:552]: missing return type

File: simple_stories_train/train_llama.py [hunk 15]
```diff
@@ -576,7 +587,7 @@ def __init__(self, filename_pattern, B, T, process_rank, num_processes):
         print0(f"DataLoader: total number of tokens: {ntok_total:,} across {len(self.files)} files")
 
         # kick things off
-        self.current_shard = None
+        self.current_shard = -1
```
Comment (danbraunai) [inline simple_stories_train/train_llama.py:590]: How come this change was made?

File: simple_stories_train/train_llama.py [hunk 16]
```diff
@@ -607,15 +618,200 @@ def next_batch(self):
         return x, y
 
 
-def print0(*args, **kwargs):
+# -----------------------------------------------------------------------------
+# Python -> C bridge utilities for saving params/grads/activations to .bin files
+
+
+def write_fp32(tensor: torch.Tensor, file: BufferedWriter):
+    t = tensor.detach().cpu().to(torch.float32)
```
Comment (danbraunai) [inline simple_stories_train/train_llama.py:626]: I think we deleted all the below from main. I'm guessing you hadn't pulled that latest main onto your branch?

File: simple_stories_train/train_llama.py [hunk 17]
```diff
@@ -607,15 +618,200 @@ def next_batch(self):
...
     # modified print that only prints from the master process
     # if this is not a distributed run, it's just a print
     if int(os.environ.get("RANK", 0)) == 0:
         print(*args, **kwargs)
 
 
 if __name__ == "__main__":
-    print0(f"Running pytorch {torch.version.__version__}")
+    import argparse
+    import time
```
Comment (danbraunai) [inline simple_stories_train/train_llama.py:812]: best to import at top of file

File: simple_stories_train/train_llama.py [hunk 18]
```diff
@@ -859,11 +1057,38 @@ def print0(*args, **kwargs):
     if args.input_val_bin:
         val_loader = DistributedDataLoader(args.input_val_bin, B, T, ddp_rank, ddp_world_size)
 
+    # -------------------------------------------------------------------------
+    # PyTorch -> C bridge: save some weights and state for C to load later as reference
+
+    # do one forward pass to generate ground truth for our C tests
```
Comment (danbraunai) [inline simple_stories_train/train_llama.py:1063]: I think we deleted all this too

**Review by danbraunai — CHANGES_REQUESTED** [review]
Thanks for this. Bunch of comments. Btw I think you may need to merge main into this branch

File: simple_stories_train/train_llama.py (hunk 1 above)
Comment (lennart-finke) [reply simple_stories_train/train_llama.py:135]: Maybe it should be Bool, yes. I got this from transformerlens (https://github.com/TransformerLensOrg/TransformerLens/blob/main/transformer_lens/components/abstract_attention.py), but did not understand why

File: simple_stories_train/train_llama.py (hunk 16 above)
Comment (lennart-finke) [reply simple_stories_train/train_llama.py:626]: Yes, looks like it.

File: simple_stories_train/train_llama.py [hunk 19]
```diff
@@ -333,10 +362,10 @@ def _init_weights(self, module: nn.Module):
 
     def forward(
         self,
-        idx: Float[torch.Tensor, "batch pos"],
-        targets: None | Float[torch.Tensor, "batch pos vocab"] = None,
+        idx: Float[Tensor, "batch pos"],
+        targets: Float[Tensor, "batch pos vocab"] | None = None,
         return_logits=True,
-    ):
+    )  -> tuple[Float[Tensor, "batch pos"] | None, Tensor | None]:
```
Comment (danbraunai) [inline simple_stories_train/train_llama.py:368]: -> `tuple[Float[Tensor, "batch pos"] | None, Float[Tensor, ""] | None]`

**Review by danbraunai — APPROVED** [review]
Looks good!
